parse/grammar/KuneiformParser.g4

```diff
@@ -0,0 +1,378 @@
+    | variable type_cast? # variable_sql_expr
+    | (table=IDENTIFIER PERIOD)? column=IDENTIFIER type_cast? # column_sql_expr
+    | sql_expr LBRACKET sql_expr RBRACKET type_cast? # array_access_sql_expr
+    | sql_expr PERIOD IDENTIFIER type_cast? # field_access_sql_expr
+    | LPAREN sql_expr RPAREN type_cast? # paren_sql_expr
+    | left=sql_expr (EQUALS | EQUATE | NEQ | LT | LTE | GT | GTE) right=sql_expr # comparison_sql_expr
+    | sql_expr NOT? IN LPAREN (sql_expr_list|select_statement) RPAREN # in_sql_expr
+    | left=sql_expr NOT? LIKE right=sql_expr # like_sql_expr
+    | <assoc=right> (NOT|PLUS|MINUS) sql_expr # unary_sql_expr
+    | element=sql_expr (NOT)? BETWEEN lower=sql_expr AND upper=sql_expr # between_sql_expr
+    | left=sql_expr IS NOT? ((DISTINCT FROM right=sql_expr) | NULL | TRUE | FALSE) # is_sql_expr
+    | sql_expr COLLATE IDENTIFIER # collate_sql_expr
+    | CASE case_clause=sql_expr?
+        (WHEN when_condition=sql_expr THEN then=sql_expr)+
+        (ELSE else_clause=sql_expr)? END            #case_expr
+    | (NOT? EXISTS)? LPAREN select_statement RPAREN type_cast? # subquery_sql_expr
+    // setting precedence for arithmetic operations:
+    | left=sql_expr CONCAT right=sql_expr # arithmetic_sql_expr
+    | left=sql_expr (STAR | DIV | MOD) right=sql_expr # arithmetic_sql_expr
+    | left=sql_expr (PLUS | MINUS) right=sql_expr # arithmetic_sql_expr
+    // setting precedence for logical operations:
+    | left=sql_expr AND right=sql_expr # logical_sql_expr
+    | left=sql_expr OR right=sql_expr # logical_sql_expr
+;
+
+sql_expr_list:
+    sql_expr (COMMA sql_expr)*
+;
+
+sql_function_call:
+    IDENTIFIER LPAREN (DISTINCT? sql_expr_list|STAR)? RPAREN #normal_call_sql
+    | IDENTIFIER LBRACKET dbid=sql_expr COMMA procedure=sql_expr RBRACKET LPAREN (sql_expr_list)? RPAREN #foreign_call_sql
+;
+
+/*
+    The following section includes parser rules for action blocks.
+*/
+// action_block is the top-level rule for an action block.
+action_block:
+    (action_statement SCOL)*
+;
+
+// action statements can only be 3 things:
+// 1. a sql statement
+// 2. a local action/procedure call.
+// 3. an extension call
+action_statement:
+    sql_statement # sql_action
+    | IDENTIFIER LPAREN (procedure_expr_list)? RPAREN # local_action
+    | (variable_list EQUALS)? IDENTIFIER PERIOD IDENTIFIER LPAREN (procedure_expr_list)? RPAREN # extension_action
+;
+
+/*
+    This section includes parser rules for procedures
+*/
+
+// procedure_block is the top-level rule for a procedure.
+procedure_block:
+    statement*
+;
+
+procedure_expr:
+    literal type_cast? # literal_procedure_expr
+    | procedure_function_call type_cast? # function_call_procedure_expr
+    | variable type_cast? # variable_procedure_expr
+    | LBRACKET (procedure_expr_list)? RBRACKET type_cast? # make_array_procedure_expr
+    | procedure_expr LBRACKET procedure_expr RBRACKET type_cast? # array_access_procedure_expr
```

Comment:
An example of using `procedure_expr` for index, which is fine, it's just not clear what `procedure_expr` we want to support can be used as index.

Reply:
Ohhhh for index in an array.

Here:

```
$arr1 text[];
$arr2 int[];
for $i in 1..100 {
  // access $i element of each array
}
```

Reply:
I guess based on current analyzer, we'll allow something like `$arr1[any_expr_return_int]`, but not sure if plpgsql support it.  
This is basically the same as the discussion here https://github.com/kwilteam/kwil-db/pull/728#discussion_r1607284055

Reply:
I'm pretty sure plpgsql does.

Reply:
Will add tests for it